### ISSUE TITLE
ROCm MI300X sum() way slower than H100

### ISSUE THREAD

**Issue by OrenLeung** [pytorch/pytorch#132964]
### 🐛 Describe the bug

even tho on `Tensor.copy_` we see major improvements on BW on MI300X compared to H100. On a similar memory BW bound op like `sum()`, we were able to achieve a read bandwidth of `3136GByte/s` on H100 SXM while only `1757.8GByte/s` on MI300X.

## Reprod
```python
import torch
from triton.testing import do_bench

x = torch.randn(2**30, device='cuda')

ms = do_bench(lambda: x.sum(dim=-1))

bandwidth_gbyte = x.numel() * x.dtype.itemsize / (10**9)

time_s = ms / 1000

bw_per_second = bandwidth_gbyte / time_s

print(bw_per_second)
```

### MI300X
latest `rocm/pytorch:rocm6.2_ubuntu22.04_py3.10_pytorch_release_2.3.0` image
```bash
sudo docker run --privileged --network=host --device=/dev/kfd --device=/dev/dri --group-add video --cap-add=SYS_PTRACE --security-opt seccomp=unconfined --ipc=host --shm-size 192G -v $(pwd):/var/lib/jenkins -it rocm/pytorch:rocm6.2_ubuntu22.04_py3.10_pytorch_release_2.3.0
```

### H100 SXM
latest `nvcr.io/nvidia/pytorch:24.07-py3` image
```bash
sudo docker run -it --ipc=host --ulimit memlock=-1 --ulimit stack=6710886 --privileged --gpus all -v $(pwd):/workspace nvcr.io/nvidia/pytorch:24.07-py3
```

cc @msaroufim @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang

**Comment by hongxiayang** [comment]
@OrenLeung Please see https://github.com/pytorch/pytorch/pull/135397 and you can verify on your end. Please let us know.

**Comment by OrenLeung** [comment]
> @OrenLeung Please see #135397 and you can verify on your end. Please let us know.

thanks for the patch @hongxiayang ! I can take a look at it once it hits nightly and will circle back on this thread with the results

**Comment by OrenLeung** [comment]
> @OrenLeung Please see #135397 and you can verify on your end. Please let us know.

hi @hongxiayang ,

I can verify the performance on the above reprod. 

Tho, when doing reducing over 1 dim on a 2d shape

H100 gets `3079.4 GByte/s` vs only MI300X `2292.7 GByte/s`. Interesting that the interesting `max_values_per_thread` only solves for 1d shape. tbh, not sure how realize the below reprod is. Let me try to find some reduce shapes from my llama3 eager training to see if there is a performance gap on reduce.

```python
import torch
from triton.testing import do_bench

x = torch.randn(2**9, 2**20, device='cuda')

ms = do_bench(lambda: x.sum(dim=-1))

bandwidth_gbyte = x.numel() * x.dtype.itemsize / (10**9)

time_s = ms / 1000

bw_per_second = bandwidth_gbyte / time_s

print(bw_per_second)
```

**Comment by jerrymannil** [comment]
@OrenLeung 
We will look into improving 2D shapes as well.
Do you have the shapes from llama3 training?

**Comment by jeffdaily** [comment]
There have been several improvements made to reduction:
- #143137
- #141397
- #141378
- #137737
